controller/AuthoringTool.php

```diff
@@ -84,56 +85,71 @@ protected function getValidatedLtiMessagePayload(): LtiMessagePayloadInterface
      */
     public function launch(): void
     {
-        $ltiMessage = $this->getLtiMessageOrRedirectToLogin();
+        try {
+            $ltiMessage = $this->getValidatedLtiMessagePayload();
 
-        $user = $this->getServiceLocator()
-            ->getContainer()
-            ->get(tao_models_classes_UserService::class)
-            ->addUser(
+            $user = $this->getUserService()->addUser(
                 $ltiMessage->getUserIdentity()->getIdentifier(),
                 helpers_Random::generateString(UserService::PASSWORD_LENGTH),
                 new core_kernel_classes_Resource(
                     $this->getAuthoringRoleService()->getValidRole($ltiMessage->getRoles())
                 )
             );
 
-        $this->getServiceLocator()
-            ->getContainer()
-            ->get(LtiService::class)
-            ->startLti1p3Session($ltiMessage, $user);
+            $this->getLtiService()->startLti1p3Session($ltiMessage, $user);
+        } catch (LtiException $exception) {
+            $this->handleLtiException($exception);
+        }
 
-        $this->forward('run', null, null, $_GET);
+        // Using a 301 Moved Permanently redirect prevents getting the request recorded in the browser's
+        // history, avoiding going back to the token validation when the user clicks the "Back" button
+        $this->redirect(_url('entry', 'Main', 'tao', $_GET), 301);
     }
 
     /**
-     * @throws ContainerExceptionInterface
      * @throws InterruptedActionException
      * @throws LtiException
-     * @throws NotFoundExceptionInterface
      */
-    private function getLtiMessageOrRedirectToLogin(): LtiMessagePayloadInterface
+    private function handleLtiException(LtiException $exception): void
     {
-        try {
-            $message = $this->getValidatedLtiMessagePayload();
-        } catch (LtiException $exception) {
-            if ($exception->getMessage() !== self::LTI_NO_MATCHING_REGISTRATION_FOUND_MESSAGE) {
-                throw $exception;
-            }
-
+        if ($exception->getMessage() === self::LTI_NO_MATCHING_REGISTRATION_FOUND_MESSAGE) {
             $this->getLogger()->warning(
                 sprintf(
                     'Missing registration for current audience. Redirecting to the login page. Exception: %s',
                     $exception
                 )
             );
-            $this->redirect(_url('login', 'Main', 'tao'));
+
+            $this->redirect(_url('login', 'Main', 'tao'));  // throws
```

Comment:
```suggestion
            $this->redirect(_url('login', 'Main', 'tao'));
```

Reply:
Can we add a return here, does not make sense to redirect follow to an exception throw. It is really confusing reading the code. 

```suggestion
            $this->redirect(_url('login', 'Main', 'tao'));
            
            return;
```